unicodetools/src/main/java/org/unicode/xml/UcdXML.java

```diff
@@ -0,0 +1,825 @@
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.xml.transform.TransformerConfigurationException;
+import org.unicode.props.IndexUnicodeProperties;
+import org.unicode.props.UcdProperty;
+import org.unicode.props.UcdPropertyValues;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.AttributesImpl;
+
+public class UcdXML {
+
+    private static final String NAMESPACE = "http://www.unicode.org/ns/2003/ucd/1.0";
+
+    private enum UCDXMLOUTPUTRANGE {
+        ALL,
+        NOUNIHAN,
+        UNIHAN;
+    }
+
+    private enum UCDXMLOUTPUTTYPE {
+        FLAT,
+        GROUPED;
+    }
+
+    private enum Range {
+        RESERVED("reserved"),
+        SURROGATE("surrogate"),
+        NONCHARACTER("noncharacter"),
+        CHARACTER("char"),
+        CJKUNIFIEDIDEOGRAPH("char"),
+        NONRANGE("nonrange");
+
+        private final String tag;
+
+        Range(String tag) {
+            this.tag = tag;
+        }
+
+        public String toString() {
+            return tag;
+        }
+    }
+
+    private static final UOption[] options = {
+        UOption.HELP_H(),
+        UOption.create("ucdversion", 'v', UOption.REQUIRES_ARG),
+        UOption.create("range", 'r', UOption.REQUIRES_ARG),
+        UOption.create("output", 'o', UOption.REQUIRES_ARG),
+        UOption.create("outputfolder", 'f', UOption.REQUIRES_ARG)
+    };
+    private static final int HELP = 0, UCDVERSION = 1, RANGE = 2, OUTPUT = 3, OUTPUTFOLDER = 4;
+
+    public static void main(String[] args) throws Exception {
+
+        VersionInfo ucdVersion = null;
+        UCDXMLOUTPUTRANGE[] ucdxmloutputranges =
+                new UCDXMLOUTPUTRANGE[] {
+                    UCDXMLOUTPUTRANGE.ALL, UCDXMLOUTPUTRANGE.NOUNIHAN, UCDXMLOUTPUTRANGE.UNIHAN
+                };
+        UCDXMLOUTPUTTYPE[] ucdxmloutputtypes =
+                new UCDXMLOUTPUTTYPE[] {UCDXMLOUTPUTTYPE.FLAT, UCDXMLOUTPUTTYPE.GROUPED};
+        File destinationFolder = null;
+
+        UOption.parseArgs(args, options);
+
+        if (options[HELP].doesOccur) {
+            System.out.println(
+                    "UcdXML --ucdversion {version number} --outputfolder {destination} "
+                            + "--range [ALL|NOUNIHAN|UNIHAN] --output [FLAT|GROUPED]");
+            System.exit(0);
+        }
+
+        try {
+            if (options[UCDVERSION].doesOccur) {
+                try {
+                    ucdVersion = VersionInfo.getInstance(options[UCDVERSION].value);
+                } catch (Exception e) {
+                    throw new IllegalArgumentException(
+                            "Could not convert "
+                                    + options[UCDVERSION].value
+                                    + " to a valid UCD version");
+                }
+            } else {
+                throw new IllegalArgumentException(
+                        "Missing command line option: --ucdversion (or -v)");
```

Comment:
how about we default to Settings.latestVersion?

and, other than maybe right now for comparison with unicode 16, do we even need a version parameter?

Reply:
Let me think about this for a bit. I see your point as the assumption is that we'll always be running on the latest version, but I'm not sure if that is a valid assumption.

Reply:
Discussed on 2/18 and implemented.